### PR TITLE
refactor: remove clone bound from entry()

### LINKED ISSUE #394: refactor: remove clone bound of key value from all interfaces

### PR DESCRIPTION
**Pull request by Xuanwo**
## What's changed and what's your intention?

> Please explain **IN DETAIL** what the changes are in this PR and why they are needed. :D

Remove the last Clone bound!

## Checklist

- [x] I have written the necessary rustdoc comments
- [x] I have added the necessary unit tests and integration tests
- [x] I have passed `make all` (or `make fast` instead if the old tests are not modified) in my local environment.

## Related issues or PRs (optional)
close #394

### CODE REVIEW

File: foyer-storage/src/generic.rs [hunk 1]
```diff
@@ -353,77 +353,69 @@ where
...
+                        Ok((key, value)) => {
+                            inner.metrics.op_bytes_get.inc_by(buf.len() as u64);
+                            Ok(Some(CachedEntry::Owned {
+                                key: Box::new(key),
+                                value: Box::new(value),
+                            }))
+                        }
+                        Err(e) => {
+                            // Remove index if the storage layer fails to get it (because of entry magic mismatch).
+                            // inner.catalog.remove(key);
```
Comment (Xuanwo) [inline foyer-storage/src/generic.rs:410]: The only place that is non-trival. Can we remove the reference of key to make this future static?

File: foyer-storage/src/storage.rs [hunk 2]
```diff
@@ -135,10 +135,10 @@ where
         Q: Hash + Eq + ?Sized;
 
     #[must_use]
-    fn get<Q>(&self, key: &Q) -> impl Future<Output = Result<Option<CachedEntry<K, V>>>> + Send
+    fn get<Q>(&self, key: &Q) -> impl Future<Output = Result<Option<CachedEntry<K, V>>>> + Send + 'static
```
Comment (Xuanwo) [inline foyer-storage/src/storage.rs:138]: `get` will be spawned on seperate runtime so it's fine for us to make it static instead. In this way, we remove the reference of `&Q` entirely and no clone anymore.

File: foyer-storage/src/generic.rs [hunk 3]
```diff
@@ -353,77 +353,69 @@ where
...
-                    .op_duration_get_miss
-                    .observe(now.elapsed().as_secs_f64());
-                return Ok(None);
-            }
-        };
-
-        match index {
-            crate::catalog::Index::Inflight { key, value } => {
-                let value = value.clone();
+        let index_value = inner.catalog.get(key);
```
Comment (Xuanwo) [inline foyer-storage/src/generic.rs:364]: Expand `&Q` by hand to make this future static.

**Comment by MrCroxx** [comment]
Thank you, @Xuanwo !! This is a very clever idea!! Let me follow your steps and apply it to foyer! 🥰

**Comment by MrCroxx** [comment]
## Benchmark

### cmd

```bash
rm -rf /p44pro/foyer && cargo build --release && RUST_BACKTRACE=1 RUST_LOG=warn ./target/release/foyer-storage-bench --dir /p44pro/foyer --capacity 204800 --file-size 64 --get-range 10000 --flushers 12 --reclaimers 12 --time 180 --writers 64 --w-rate 8 --ticket-insert-rate-limit 500 --readers 1024 --r-rate 1 --runtime --metrics --compression none --distribution zipf --distribution-zipf-s 0.8
```

### main (as https://github.com/MrCroxx/foyer/pull/397#issuecomment-2068661620)

```
Total:
disk total iops: 7969.2
disk total throughput: 527.7 MiB/s
disk read iops: 1006.3
disk read throughput: 66.8 MiB/s
disk write iops: 6963.0
disk write throughput: 460.9 MiB/s
insert iops: 6879.6/s
insert throughput: 430.0 MiB/s
insert lat p50: 38us
insert lat p90: 73us
insert lat p99: 385us
insert lat p999: 871us
insert lat p9999: 1711us
insert lat p99999: 2655us
insert lat pmax: 3199us
get iops: 16335.5/s
get miss: 0.82%
get throughput: 1.0 GiB/s
get hit lat p50: 891us
get hit lat p90: 1991us
get hit lat p99: 2975us
get hit lat p999: 6271us
get hit lat p9999: 8511us
get hit lat p99999: 9087us
get hit lat pmax: 9343us
get miss lat p50: 125us
get miss lat p90: 347us
get miss lat p99: 767us
get miss lat p999: 1431us
get miss lat p9999: 2335us
get miss lat p99999: 2463us
get miss lat pmax: 2463us
```

### PR

```
Total:
disk total iops: 7879.6
disk total throughput: 522.0 MiB/s
disk read iops: 947.5
disk read throughput: 62.9 MiB/s
disk write iops: 6932.1
disk write throughput: 459.1 MiB/s
insert iops: 6854.8/s
insert throughput: 428.4 MiB/s
insert lat p50: 37us
insert lat p90: 72us
insert lat p99: 351us
insert lat p999: 843us
insert lat p9999: 1551us
insert lat p99999: 2431us
insert lat pmax: 2847us
get iops: 16379.8/s
get miss: 1.09%
get throughput: 1.0 GiB/s
get hit lat p50: 839us
get hit lat p90: 1967us
get hit lat p99: 2847us
get hit lat p999: 3295us
get hit lat p9999: 8159us
get hit lat p99999: 9215us
get hit lat pmax: 9727us
get miss lat p50: 137us
get miss lat p90: 341us
get miss lat p99: 699us
get miss lat p999: 1287us
get miss lat p9999: 2415us
get miss lat p99999: 2831us
get miss lat pmax: 2831us
```

**Pull request merged**